parquet-column/src/main/java/org/apache/parquet/column/ColumnProperty.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.parquet.column;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import org.apache.parquet.hadoop.metadata.ColumnPath;
+
+/**
+ * Represents a Parquet property that may have different values for the different columns.
+ */
+abstract class ColumnProperty<T> {
+  private static class DefaultColumnProperty<T> extends ColumnProperty<T> {
+    private final T defaultValue;
+
+    private DefaultColumnProperty(T defaultValue) {
+      this.defaultValue = defaultValue;
+    }
+
+    @Override
+    public T getDefaultValue() {
+      return defaultValue;
+    }
+
+    @Override
+    public T getValue(ColumnPath columnPath) {
+      return getDefaultValue();
+    }
+
+    @Override
+    public String toString() {
+      return Objects.toString(getDefaultValue());
+    }
+  }
+
+  private static class MultipleColumnProperty<T> extends DefaultColumnProperty<T> {
+    private final Map<ColumnPath, T> values;
+
+    private MultipleColumnProperty(T defaultValue, Map<ColumnPath, T> values) {
+      super(defaultValue);
+      assert !values.isEmpty();
```

Comment:
Is it better to use a warning here?

Reply:
I usually use `assert` to ensure something that if not happens then it is a coding issue. Here, this private constructor is invoked only at one place where the map is not empty. It is just that I've coded this line first and wanted to not forget about the check :)
(`assert` is good for such checks because they do not impact performance in production.)

Reply:
I see, thanks for the explanation.